src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1321,9 +1367,10 @@ ObjectMonitor* ObjectSynchronizer::inflate(Thread* current, oop object,
       ObjectMonitor* inf = mark.monitor();
       markWord dmw = inf->header();
       assert(dmw.is_neutral(), "invariant: header=" INTPTR_FORMAT, dmw.value());
-      if (LockingMode == LM_LIGHTWEIGHT && inf->is_owner_anonymous() && is_lock_owned(current, object)) {
-        inf->set_owner_from_anonymous(current);
-        JavaThread::cast(current)->lock_stack().remove(object);
+      if (LockingMode == LM_LIGHTWEIGHT && inf->is_owner_anonymous() &&
+          thread != nullptr && thread->lock_stack().contains(object)) {
+        inf->set_owner_from_anonymous(thread);
```

Comment:
I thought one of the invariants of lightweight locking was that a transition from anonymously owned to actual owner, can only be performed by the actual owner?

That said, can we actually reach here if not with the current thread? If we are re-locking for deopt then, IIUC, the object has not yet escaped, and has to be unlocked and not with an inflated monitor. In such a case we can't be racing with hashcode inflation either. But perhaps there are other circumstances where re-locking is needed?

Reply:
Unsure if the `relock_objects` code only re-locks escaped and elided locks, or all elided locks. I believe it is the second. And if we re-lock nested elided locks then they can very well have escaped. (But the lock is guaranteed to be owned by the `deoptee_thread`)

Reply:
Given the extra test that @reinrich provided that is exactly what occurs. A nested eliminated lock is re-locked. But the lock is already escaped and contended by another thread.

Reply:
> Given the extra test that @reinrich provided that is exactly what occurs. A nested eliminated lock is re-locked. But the lock is already escaped and contended by another thread.

We didn't reach there with LM_LIGHTWEIGHT. The lock got inflated during warm-up because of the recursive monitorenter in the interpreter. The owner was set to the current thread then. I've improved the test to use a new instance after warm-up. If you want you can cherry-pick here: https://github.com/reinrich/jdk/commit/c28181d585e5ceeaa34e7e6afb4311d59cf990b2

Reply:
> I thought one of the invariants of lightweight locking was that a transition from anonymously owned to actual owner, can only be performed by the actual owner?

To follow up to my own query, this is not quite the case. We introduced the notion of being anonymously owned because under normal conditions if a thread B tries to lock a monitor fast-locked by thread A, it has to inflate the monitor, but it does not know who the owning thread is. Hence it sets it to anonymous. Then when thread A goes to release the monitor, it seems it is inflated, and anonymously owned and so sets the owner to itself, then unlocks.

So as long as we know who the owner thread is, then it should be safe to update the owner from being anonymous.

Also thanks for clarifying the nested lock elimination case, which can lead to the inflation case.